src/webui/src/components/trial-detail/TableList.tsx

```diff
@@ -567,87 +553,99 @@ class TableList extends React.Component<TableListProps, TableListState> {
             this.setState({
                 tableColumns: this.initTableColumnList(columnList),
                 allColumnList: this.getAllColumnKeys()
-                }, () => {this.updateData();
-            });
+                }, () => {
+                    this.updateData();
+                });
         }
     }
 
+    // slice all table data into current page data
     updateData(): void {
-        const tableSource: Array<TableRecord> = JSON.parse(JSON.stringify(this.props.tableSource));
+        const tableSource: Array<TableRecord> = this.props.tableSource;
+        const { offset, perPage, sortMessage } = this.state;
+
+        if (sortMessage.field !== '') {	
+            tableSource.sort(function (a, b): any {	
+                if (a[sortMessage.field] === undefined) {	
+                    return 1;	
+                }	
+                if (b[sortMessage.field] === undefined) {	
+                    return -1;	
+                }	
+                return (sortMessage.isDescend ? a[sortMessage.field] < b[sortMessage.field] : a[sortMessage.field] > b[sortMessage.field]) ? 1 : -1;	
+            });	
+        }
 
-        const tableSlice = tableSource.slice(this.state.offset, this.state.offset + this.state.perPage)
-        
+        const tableSlice = tableSource.slice(offset, offset + perPage)
+        const curPageCount = Math.ceil(tableSource.length / perPage)
         this.setState({
-            tableSourceForSort: tableSlice,
-            pageCount: Math.ceil(tableSource.length / this.state.perPage),
+            tablePerPage: tableSlice,
+            pageCount: curPageCount,
         });
     }
     
+    // update data when click the page index of pagination
     handlePageClick = (evt: any): void => {
         const selectedPage = evt.selected;
         const offset = selectedPage * this.state.perPage;
         
         this.setState({ 
             currentPage: selectedPage, 
-            offset: offset }, 
-            () => { this.updateData();
+            offset: offset
+        }, () => {
+            this.updateData();
         });
     }
 
-    updateperPage = (event: React.FormEvent<HTMLDivElement>, item: IDropdownOption | undefined): void => {
-        // clear input value and re-render table
+    // update per page items when click the dropdown of pagination
+    updatePerPage = (event: React.FormEvent<HTMLDivElement>, item: IDropdownOption | undefined): void => {
+        const { pageCount } = this.state;
+        
         if (item !== undefined) {
+            const currentPerPage = item.key === 'all' ? this.props.tableSource.length: Number(item.key)
+            const currentPageCount = this.props.tableSource.length <= currentPerPage ? 1 : pageCount
+            
             this.setState({ 
-                perPage: item.key === 'all' ? this.props.tableSource.length: Number(item.key) },
-                () => {this.updateData();
+                perPage: currentPerPage,
+                offset: 0,
```

Comment:
so, when ` item.key ` changed, `offset ` always need to be setted zero?

Reply:
Yes, reset to the first page refer to pai's implementation. If the number of total data is 30, currently on page 3, select from 10 to 20 items would out of the data range.